In [62]:
import pandas as pd
import numpy as np

In [63]:
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/hepatitis/hepatitis.data')


In [64]:
df.head()

2  30  2.1  1  2.2 2.3 2.4 2.5 1.1 2.6 2.7 2.8 2.9 2.10  1.00   85   18  \
0  2  50    1  1    2   1   2   2   1   2   2   2   2    2  0.90  135   42   
1  2  78    1  2    2   1   2   2   2   2   2   2   2    2  0.70   96   32   
2  2  31    1  ?    1   2   2   2   2   2   2   2   2    2  0.70   46   52   
3  2  34    1  2    2   2   2   2   2   2   2   2   2    2  1.00    ?  200   
4  2  34    1  2    2   2   2   2   2   2   2   2   2    2  0.90   95   28   

   4.0   ?  1.2  
0  3.5   ?    1  
1  4.0   ?    1  
2  4.0  80    1  
3  4.0   ?    1  
4  4.0  75    1

In [55]:
X=df.replace('?', '2')
df.dtypes

Class               int64
AGE                 int64
SEX                 int64
STEROID            object
ANTIVIRALS          int64
FATIGUE            object
MALAISE            object
ANOREXIA           object
LIVER BIG          object
LIVER FIRM         object
SPLEEN PALPABLE    object
SPIDERS            object
ASCITES            object
VARICES            object
BILIRUBIN          object
ALK PHOSPHATE      object
SGOT               object
ALBUMIN            object
PROTIME            object
HISTOLOGY           int64
dtype: object

In [65]:
df.isnull().sum()

2       0
30      0
2.1     0
1       0
2.2     0
2.3     0
2.4     0
2.5     0
1.1     0
2.6     0
2.7     0
2.8     0
2.9     0
2.10    0
1.00    0
85      0
18      0
4.0     0
?       0
1.2     0
dtype: int64

In [54]:
X.columns =['Class','AGE','SEX','STEROID','ANTIVIRALS','FATIGUE','MALAISE','ANOREXIA','LIVER BIG','LIVER FIRM','SPLEEN PALPABLE','SPIDERS','ASCITES','VARICES','BILIRUBIN','ALK PHOSPHATE','SGOT','ALBUMIN','PROTIME','HISTOLOGY']

In [37]:
X.head()

Class  AGE  SEX STEROID  ANTIVIRALS FATIGUE MALAISE ANOREXIA LIVER BIG  \
0      2   50    1       1           2       1       2        2         1   
1      2   78    1       2           2       1       2        2         2   
2      2   31    1       1           1       2       2        2         2   
3      2   34    1       2           2       2       2        2         2   
4      2   34    1       2           2       2       2        2         2   

  LIVER FIRM SPLEEN PALPABLE SPIDERS ASCITES VARICES BILIRUBIN ALK PHOSPHATE  \
0          2               2       2       2       2      0.90           135   
1          2               2       2       2       2      0.70            96   
2          2               2       2       2       2      0.70            46   
3          2               2       2       2       2      1.00             1   
4          2               2       2       2       2      0.90            95   

  SGOT ALBUMIN PROTIME  HISTOLOGY  
0   42     3.5       1          1  
1   32     4.0       1          1  
2   52     4.0      80          1  
3  200     4.0       1          1  
4   28     4.0      75          1

#### 1. Class: 1-DIE, 2-LIVE 
     2. AGE: 10, 20, 30, 40, 50, 60, 70, 80
     3. SEX: male, female
     4. STEROID: no, yes
     5. ANTIVIRALS: no, yes
     6. FATIGUE: no, yes
     7. MALAISE: no, yes
     8. ANOREXIA: no, yes
     9. LIVER BIG: no, yes
    10. LIVER FIRM: no, yes
    11. SPLEEN PALPABLE: no, yes
    12. SPIDERS: no, yes
    13. ASCITES: no, yes
    14. VARICES: no, yes
    15. BILIRUBIN: 0.39, 0.80, 1.20, 2.00, 3.00, 4.00
        -- see the note below
    16. ALK PHOSPHATE: 33, 80, 120, 160, 200, 250
    17. SGOT: 13, 100, 200, 300, 400, 500, 
    18. ALBUMIN: 2.1, 3.0, 3.8, 4.5, 5.0, 6.0
    19. PROTIME: 10, 20, 30, 40, 50, 60, 70, 80, 90
    20. HISTOLOGY: no, yes

from sklearn.preprocessing import Imputer

imputer=Imputer(missing_values="NaN",strategy='mean',axis=1)

imputer=imputer.fit(df)

X=imputer.transform(df)

X=X.astype('int64')

In [28]:
df=X
df.head()

Class  AGE  SEX STEROID  ANTIVIRALS FATIGUE MALAISE ANOREXIA LIVER BIG  \
0      2   50    1       1           2       1       2        2         1   
1      2   78    1       2           2       1       2        2         2   
2      2   31    1       1           1       2       2        2         2   
3      2   34    1       2           2       2       2        2         2   
4      2   34    1       2           2       2       2        2         2   

  LIVER FIRM SPLEEN PALPABLE SPIDERS ASCITES VARICES BILIRUBIN ALK PHOSPHATE  \
0          2               2       2       2       2      0.90           135   
1          2               2       2       2       2      0.70            96   
2          2               2       2       2       2      0.70            46   
3          2               2       2       2       2      1.00             1   
4          2               2       2       2       2      0.90            95   

  SGOT ALBUMIN PROTIME  HISTOLOGY  
0   42     3.5       1          1  
1   32     4.0       1          1  
2   52     4.0      80          1  
3  200     4.0       1          1  
4   28     4.0      75          1

X = df.drop(columns=['Class'])

In [29]:
X.head()

Class  AGE  SEX STEROID  ANTIVIRALS FATIGUE MALAISE ANOREXIA LIVER BIG  \
0      2   50    1       1           2       1       2        2         1   
1      2   78    1       2           2       1       2        2         2   
2      2   31    1       1           1       2       2        2         2   
3      2   34    1       2           2       2       2        2         2   
4      2   34    1       2           2       2       2        2         2   

  LIVER FIRM SPLEEN PALPABLE SPIDERS ASCITES VARICES BILIRUBIN ALK PHOSPHATE  \
0          2               2       2       2       2      0.90           135   
1          2               2       2       2       2      0.70            96   
2          2               2       2       2       2      0.70            46   
3          2               2       2       2       2      1.00             1   
4          2               2       2       2       2      0.90            95   

  SGOT ALBUMIN PROTIME  HISTOLOGY  
0   42     3.5       1          1  
1   32     4.0       1          1  
2   52     4.0      80          1  
3  200     4.0       1          1  
4   28     4.0      75          1

In [30]:
y = df['Class'].values

In [31]:
y[::]

array([2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 1, 1, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       1, 2, 2, 2, 1, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1,
       2, 2, 1, 2, 2, 1, 2, 2, 2, 1, 2, 1, 2, 2, 2, 1, 2, 1, 2, 2, 1, 2,
       1, 2, 2, 2, 2, 2, 2, 1, 2, 1, 2, 2, 2, 2, 2, 2, 1, 2, 1, 2, 1, 2,
       2, 1, 2, 2, 2, 1, 2, 2, 1, 2, 1, 1, 2, 1, 1, 2, 2, 1, 2, 2, 2, 1],
      dtype=int64)

In [32]:
from sklearn.model_selection import train_test_split
#split dataset into train and test data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1, stratify=y)

In [39]:
X_train.shape

(123, 20)

In [41]:
X_test.shape

(31, 20)

In [34]:
from sklearn.neighbors import KNeighborsClassifier
# Create KNN classifier
knn = KNeighborsClassifier(n_neighbors = 10)

knn.fit(X_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=10, p=2,
           weights='uniform')

In [48]:
y_pred=list(knn.predict(X_test))

In [49]:
knn.score(X_test, y_test)

0.7741935483870968

In [ ]:
acc